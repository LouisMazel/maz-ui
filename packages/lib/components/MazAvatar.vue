<template>
  <Component
    :is="componentType"
    :style="{ fontSize: size }"
    class="m-avatar"
    :class="[
      {
        '--has-link': isLink,
      },
    ]"
    :href="href"
    :to="to"
    :target="isLink ? target : undefined"
  >
    <div
      class="m-avatar__wrapper"
      :tabindex="clickable ? 0 : -1"
      :class="{
        '--has-shadow': !noElevation,
        '--bordered': bordered,
        '--clickable': clickable,
        '--square': square,
        '--has-initial': !src && caption,
      }"
    >
      <MazLazyImg
        v-if="src"
        class="m-avatar__picture maz-max-w-full"
        :image="src"
        :alt="alt"
        image-height-full
        :no-loader="noLoader"
        @click="clickable ? $emit('click', $event) : null"
      />
      <slot v-if="caption && !src" name="round-text">
        <span class="m-avatar__initial"> {{ getInitials(caption) }} </span>
      </slot>

      <button
        v-if="clickable"
        type="button"
        tabindex="-1"
        class="m-avatar__button"
        :style="{
          backgroundColor: src
            ? `var(--maz-color-${buttonColor}-alpha)`
            : `var(--maz-color-${buttonColor})`,
        }"
        @click="$emit('click', $event)"
      >
        <slot v-if="!noClickableIcon" name="icon">
          <PencilIcon class="m-avatar__button__icon" />
        </slot>
      </button>
    </div>
    <slot name="caption">
      <p v-if="showCaption && caption" class="m-avatar__caption">
        {{ caption }}
      </p>
    </slot>
  </Component>
</template>

<script lang="ts" setup>
  import { computed, defineAsyncComponent, type PropType } from 'vue'

  import type { Color } from './types'
  export type { Color }

  const MazLazyImg = defineAsyncComponent(() => import('./MazLazyImg.vue'))
  const PencilIcon = defineAsyncComponent(() => import('./../icons/pencil.svg'))

  const props = defineProps({
    src: {
      type: String as PropType<undefined | null | string>,
      default: undefined,
    },
    caption: {
      type: String as PropType<undefined | null | string>,
      default: undefined,
    },
    href: { type: String, default: undefined },
    to: { type: Object, default: undefined },
    alt: { type: String, default: 'avatar image' },
    target: { type: String, default: '_self' },
    size: { type: String, default: undefined },
    bordered: { type: Boolean, default: false },
    clickable: { type: Boolean, default: false },
    square: { type: Boolean, default: false },
    noElevation: { type: Boolean, default: false },
    showCaption: { type: Boolean, default: false },
    imageHeightFull: { type: Boolean, default: false },
    noLoader: { type: Boolean, default: false },
    buttonColor: {
      type: String as PropType<Color>,
      default: 'info',
    },
    /** Remove the icon on hover when component is clickable */
    noClickableIcon: { type: Boolean, default: false },
    letterCount: { type: Number, default: undefined },
  })

  const componentType = computed(() => (props.to ? 'RouterLink' : props.href ? 'a' : 'div'))
  const isLink = computed(() => !!props.to || !!props.href)

  function getInitials(name: string, lettersCount = props.letterCount) {
    const words = name.split(' ')

    const initials = words.map((word) => word[0])

    const letters = initials.join('')

    return letters.slice(0, lettersCount)
  }

  defineEmits(['click'])
</script>

<style lang="postcss" scoped>
  .m-avatar {
    @apply maz-inline-flex maz-flex-col maz-flex-center;
    @apply maz-no-underline !important;

    &__caption {
      @apply maz-mt-2 maz-w-full maz-truncate maz-text-center maz-font-medium maz-capitalize;
    }

    &__initial {
      @apply maz-capitalize maz-text-white;

      font-size: 1.5em;
    }

    &__wrapper {
      @apply maz-relative maz-flex maz-flex-none maz-justify-center maz-overflow-hidden
        maz-rounded-full maz-bg-color-lighter;

      height: 3em;
      width: 3em;

      &.--clickable {
        & .m-avatar__button {
          @apply maz-absolute maz-inset-0 maz-flex maz-w-full
            maz-cursor-pointer maz-rounded maz-border-none maz-bg-transparent
            maz-opacity-0 maz-outline-none maz-transition-all maz-duration-200 maz-flex-center;

          transform: scale(0);

          &__icon {
            @apply maz-text-white;
          }
        }

        &:not(.--square) .m-avatar__button {
          @apply maz-rounded-full;
        }

        &:hover,
        &:focus {
          & .m-avatar__picture {
            filter: blur(1.5px);
          }

          & .m-avatar__button {
            @apply maz-opacity-100;

            transform: scale(1.05);
          }
        }
      }

      &.--bordered {
        @apply maz-border maz-border-solid maz-border-white;
      }

      &.--square {
        @apply maz-rounded;
      }

      &.--has-shadow {
        @apply maz-shadow;
      }

      &.--has-initial {
        @apply maz-items-center maz-bg-primary;
      }
    }

    &.--has-link {
      @apply maz-cursor-pointer;
    }
  }
</style>
