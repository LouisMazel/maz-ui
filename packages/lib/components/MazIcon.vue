<template>
  <!-- eslint-disable vue/no-v-html -->
  <!-- eslint-disable vue/html-self-closing -->
  <svg
    v-if="svgElSource"
    ref="svgElem"
    width="1em"
    height="1em"
    v-bind="{
      ...getSvgAttrs(svgElSource),
      ...filterAttrs($attrs),
    }"
    :style="`font-size: ${size}`"
    v-html="getSvgContent(svgElSource)"
  ></svg>
  <!-- eslint-enable vue/no-v-html -->
  <!-- eslint-enable vue/html-self-closing -->
</template>

<script lang="ts" setup>
  import { nextTick, onMounted, ref, computed, watchEffect, type PropType } from 'vue'

  import { injectStrict } from '../modules/helpers/inject-strict'

  const cache: Record<string, Promise<SVGElement>> = {}
  const svgElSource = ref<SVGElement>()
  const svgElem = ref<SVGElement>()

  const getMazIconPath = () => {
    try {
      return injectStrict<string>('mazIconPath')
    } catch {
      return undefined
    }
  }

  const props = defineProps({
    src: { type: String, default: undefined },
    path: { type: String, default: undefined },
    name: { type: String, default: undefined },
    size: { type: String, default: undefined },
    title: { type: String, default: undefined },
    transformSource: {
      type: Function as PropType<(param: SVGElement) => typeof param>,
      default: (svg: SVGElement) => svg,
    },
  })

  const emits = defineEmits(['loaded', 'unloaded', 'error'])

  const iconPath = computed(() => props.path ?? getMazIconPath())
  const fullSrc = computed(() => {
    if (props.src) {
      return props.src
    } else if (iconPath.value) {
      return `${iconPath.value}/${props.name}.svg`
    } else {
      return `/${props.name}.svg`
    }
  })

  onMounted(() => {
    if (!props.name && !props.src) {
      console.error('[maz-ui](MazIcon) you should provide "name" or "src" as prop')
    }
  })

  const setTitle = (svg: SVGElement, title: string) => {
    const titleTags = svg.querySelectorAll('title')
    if (titleTags.length > 0) {
      // overwrite existing title
      titleTags[0].textContent = title
    } else {
      // create a title element if one doesn't already exist
      const titleEl = document.createElementNS('http://www.w3.org/2000/svg', 'title')
      titleEl.textContent = title
      svg.append(titleEl)
    }
  }

  const filterAttrs = (attrs: Record<string, unknown>) => {
    return Object.keys(attrs).reduce((result, key) => {
      if (attrs[key] !== false && attrs[key] !== null && attrs[key] !== undefined) {
        result[key] = attrs[key]
      }
      return result
    }, {})
  }

  const getSvgAttrs = (svgEl: SVGElement) => {
    // copy attrs
    const svgAttrs = {}
    const attrs = svgEl.attributes
    if (!attrs) {
      return svgAttrs
    }
    for (let i = attrs.length - 1; i >= 0; i--) {
      svgAttrs[attrs[i].name] = attrs[i].value
    }
    return svgAttrs
  }

  const getSvgContent = (svgEl: SVGElement) => {
    let svgElNode = svgEl.cloneNode(true)
    svgElNode = props.transformSource(svgEl)

    if (props.title) {
      setTitle(svgElNode as SVGElement, props.title)
    }

    // copy inner html
    return svgEl.innerHTML
  }

  const getSource = async (src: string) => {
    // fill cache by src with promise
    if (!cache[src]) {
      // download
      cache[src] = download(src)
    }

    // inline svg when cached promise resolves
    try {
      svgElSource.value = await cache[src]
      // wait to render
      await nextTick()
      emits('loaded', svgElem.value)
    } catch (error) {
      if (svgElSource.value) {
        svgElSource.value = undefined
        emits('unloaded')
      }
      // remove cached rejected promise so next image can try load again
      delete cache[src]
      emits('error', error)
    }
  }

  const download = (url: string): Promise<SVGElement> => {
    return new Promise((resolve, reject) => {
      const request = new XMLHttpRequest()
      request.open('GET', url, true)

      request.addEventListener('load', () => {
        if (request.status >= 200 && request.status < 400) {
          try {
            // Setup a parser to convert the response to text/xml in order for it to be manipulated and changed
            const parser = new DOMParser()
            const result = parser.parseFromString(request.responseText, 'text/xml')
            let svgEl = result.querySelectorAll('svg')[0] as SVGElement
            if (svgEl) {
              svgEl = props.transformSource(svgEl)
              resolve(svgEl)
            } else {
              reject(new Error('Loaded file is not valid SVG"'))
            }
          } catch (error) {
            reject(error)
          }
        } else {
          reject(new Error('Error loading SVG'))
        }
      })

      request.addEventListener('error', () => reject())
      request.send()
    })
  }

  watchEffect(() => getSource(fullSrc.value))
</script>
