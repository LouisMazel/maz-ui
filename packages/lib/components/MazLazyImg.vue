<script lang="ts" setup>
import type { vLazyImgOptions } from './../modules/directives/v-lazy-img/types'
import type { Image } from './types'
import { computed, defineAsyncComponent, type HTMLAttributes } from 'vue'

import { vLazyImg } from './../modules/directives/v-lazy-img/lazy-img.directive'

export type { Image }

defineOptions({
  inheritAttrs: false,
})

const props = withDefaults(defineProps<Props>(), {
  style: undefined,
  class: undefined,
  image: undefined,
  src: undefined,
  alt: undefined,
  observerOptions: undefined,
  fallbackSrc: undefined,
})

defineEmits<{
  /** Emitted when the image is intersecting */
  (name: 'intersecting', el: Element): void
  /** Emitted when the image is loading */
  (name: 'loading', el: Element): void
  /** Emitted when the image is loaded */
  (name: 'loaded', el: Element): void
  /** Emitted when the image is in error */
  (name: 'error', el: Element): void
}>()

const MazSpinner = defineAsyncComponent(() => import('./MazSpinner.vue'))

export interface Props {
  /** The style of the component */
  style?: HTMLAttributes['style']
  /** The class of the component */
  class?: HTMLAttributes['class']
  /** @deprecated Use `src` instead */
  image?: Image | null
  /**
   * The source of the image
   * @type {string | Image | null}
   */
  src?: Image | null
  /** The alt of the image */
  alt?: string
  /** Display the fallback image */
  noPhoto?: boolean
  /** Remove the loader */
  noLoader?: boolean
  /** Remove the observer once the image is loaded */
  noObserverOnce?: boolean
  /** Remove the observer once the image is loaded */
  loadOnce?: boolean
  /** Make the image height full */
  imageHeightFull?: boolean
  /** The options of the observer */
  observerOptions?: vLazyImgOptions['observerOptions']
  /** The fallback src to replace the src on loading error */
  fallbackSrc?: string
  /** The classes of the image element */
  imgClass?: HTMLAttributes['class']
  /** The image will be displayed in full width */
  block?: boolean
}

const src = computed(() => props.image || props.src)

const sources = computed(() => {
  return typeof src.value === 'string' ? [{ srcset: src.value }] : src.value?.sources
})
</script>

<template>
  <picture
    v-lazy-img="{
      noPhoto,
      loadOnce,
      observerOptions,
      fallbackSrc,
      observerOnce: !noObserverOnce,
      onIntersecting: (el) => $emit('intersecting', el),
      onLoading: (el) => $emit('loading', el),
      onLoaded: (el) => $emit('loaded', el),
      onError: (el) => $emit('error', el),
    }"
    class="m-lazy-img-component"
    :class="[{ '--use-loader': !noLoader, '--height-full': imageHeightFull, '--block': block }, props.class]"
    :style
  >
    <source
      v-for="({ srcset, media }, sourceIndex) in sources"
      :key="sourceIndex"
      :data-lazy-srcset="srcset"
      :media
    >
    <img
      v-bind="$attrs"
      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
      loading="lazy"
      :alt
      :class="imgClass"
    >
    <div v-if="!noLoader" class="m-lazy-img-component-loader">
      <MazSpinner size="2em" />
    </div>
    <slot />
  </picture>
</template>

<style lang="postcss" scoped>
  .m-lazy-img-component {
    @apply maz-relative maz-inline-flex maz-align-top maz-flex-center;

    &.--block {
      @apply maz-w-full;

      img {
        @apply maz-w-full;
      }
    }

    &-loader {
      @apply maz-absolute maz-inset-0 maz-hidden maz-flex-center;
    }

    /* &:not(.m-lazy-error, .m-lazy-no-photo) img {
      @apply maz-h-full maz-w-full;
    } */

    &.--height-full img {
      @apply maz-max-h-full maz-w-min maz-max-w-min !important;
    }

    &.m-lazy-error:not(.m-lazy-no-photo) {
      @apply maz-bg-color-light;
      img {
        @apply maz-h-1/2 maz-w-1/2;
      }
    }

    &.m-lazy-loading {
      & .m-lazy-img-component-loader {
        @apply maz-flex;
      }
    }
  }
</style>
