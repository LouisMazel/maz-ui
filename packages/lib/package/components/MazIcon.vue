<template>
  <!-- eslint-disable vue/no-v-html -->
  <svg
    v-if="svgElSource"
    ref="svgElem"
    width="1em"
    height="1em"
    v-bind="{
      ...getSvgAttrs(svgElSource),
      ...filterAttrs($attrs),
    }"
    :style="`font-size: ${size}`"
    v-html="getSvgContent(svgElSource)"
  ></svg>
  <!-- eslint-enable vue/no-v-html -->
</template>

<!-- // inheritAttrs: false -->
<script lang="ts" setup>
  import { nextTick, onMounted, PropType, ref, watch, computed } from 'vue'
  import { injectStrict } from '../utils/injectStrict'

  const cache: Record<string, Promise<SVGElement>> = {}
  const svgElSource = ref<SVGElement>()
  const svgElem = ref<SVGElement>()

  const mazIconPath = injectStrict<string>('mazIconPath')

  const props = defineProps({
    src: { type: String, default: undefined },
    path: { type: String, default: undefined },
    name: { type: String, default: undefined },
    size: { type: String, default: '1.5rem' },
    title: { type: String, default: undefined },
    transformSource: {
      type: Function as PropType<(param: SVGElement) => typeof param>,
      default: (svg: SVGElement) => svg,
    },
    keepDuringLoading: { type: Boolean, default: true },
  })

  const emits = defineEmits(['loaded', 'unloaded', 'error'])

  const iconPath = computed(() => props.path ?? mazIconPath)
  const fullSrc = computed(() =>
    props.src
      ? props.src
      : iconPath.value
      ? `${iconPath.value}/${props.name}.svg`
      : `/${props.name}.svg`,
  )

  onMounted(() => {
    if (!props.name && !props.src) {
      throw new Error('[MazIcon] you should provide "name" or "src" as prop')
    }
    getSource(fullSrc.value)
  })

  const setTitle = (svg: SVGElement, title: string) => {
    const titleTags = svg.getElementsByTagName('title')
    if (titleTags.length) {
      // overwrite existing title
      titleTags[0].textContent = title
    } else {
      // create a title element if one doesn't already exist
      const titleEl = document.createElementNS(
        'http://www.w3.org/2000/svg',
        'title',
      )
      titleEl.textContent = title
      svg.appendChild(titleEl)
    }
  }

  const filterAttrs = (attrs: Record<string, unknown>) => {
    return Object.keys(attrs).reduce((result, key) => {
      if (
        attrs[key] !== false &&
        attrs[key] !== null &&
        attrs[key] !== undefined
      ) {
        result[key] = attrs[key]
      }
      return result
    }, {})
  }

  const getSvgAttrs = (svgEl: SVGElement) => {
    // copy attrs
    const svgAttrs = {}
    const attrs = svgEl.attributes
    if (!attrs) {
      return svgAttrs
    }
    for (let i = attrs.length - 1; i >= 0; i--) {
      svgAttrs[attrs[i].name] = attrs[i].value
    }
    return svgAttrs
  }

  const getSvgContent = (svgEl: SVGElement) => {
    let svgElNode = svgEl.cloneNode(true)
    svgElNode = props.transformSource(svgEl)

    if (props.title) {
      setTitle(svgElNode as SVGElement, props.title)
    }

    // copy inner html
    return svgEl.innerHTML
  }

  const getSource = async (src: string) => {
    // fill cache by src with promise
    if (!cache[src]) {
      // download
      cache[src] = download(src)
    }

    // inline svg when cached promise resolves
    try {
      const svg = await cache[src]
      svgElSource.value = svg
      // wait to render
      await nextTick()
      emits('loaded', svgElem.value)
    } catch (err) {
      if (svgElSource.value) {
        svgElSource.value = undefined
        emits('unloaded')
      }
      // remove cached rejected promise so next image can try load again
      delete cache[src]
      emits('error', err)
    }
  }

  const download = (url: string): Promise<SVGElement> => {
    return new Promise((resolve, reject) => {
      const request = new XMLHttpRequest()
      request.open('GET', url, true)

      request.onload = () => {
        if (request.status >= 200 && request.status < 400) {
          try {
            // Setup a parser to convert the response to text/xml in order for it to be manipulated and changed
            const parser = new DOMParser()
            const result = parser.parseFromString(
              request.responseText,
              'text/xml',
            )
            const svgEl = result.getElementsByTagName('svg')[0]
            if (svgEl) {
              // svgEl = this.transformSource(svgEl);
              resolve(svgEl)
            } else {
              reject(new Error('Loaded file is not valid SVG"'))
            }
          } catch (e) {
            reject(e)
          }
        } else {
          reject(new Error('Error loading SVG'))
        }
      }

      request.onerror = reject
      request.send()
    })
  }

  watch(
    () => props.src,
    () => getSource(fullSrc.value),
  )
</script>
