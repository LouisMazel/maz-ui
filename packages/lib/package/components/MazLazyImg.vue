<template>
  <picture
    v-lazy-img="{
      noPhoto: noPhoto,
      observerOnce: !noObserverOnce,
      loadOnce: loadOnce,
      onIntersecting: (el: HTMLElement) => $emit('intersecting', el),
      onLoading: (el: HTMLElement) => $emit('loading', el),
      onLoaded: (el: HTMLElement) => $emit('loaded', el),
      onError: (el: HTMLElement) => $emit('error', el),
      observerOptions: observerOptions,
    }"
    class="m-lazy-img-component"
    :class="{ '-use-loader': !noLoader, '--height-full': imageHeightFull }"
  >
    <source
      v-for="({ srcset, media }, sourceIndex) in sources"
      :key="sourceIndex"
      :data-srcset="srcset"
      :media="media"
    />
    <img
      v-bind="$attrs"
      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
    />
    <div v-if="!noLoader" class="m-lazy-img-component-loader">
      <MazSpinner color="black" size="2em" />
    </div>
    <slot />
  </picture>
</template>

<script lang="ts">
  export type { Image } from './types'
</script>

<script lang="ts" setup>
  import { Prop, computed } from 'vue'
  import { vLazyImgOptions, vLazyImg } from '../directives/v-lazy-img'

  import MazSpinner from './MazSpinner.vue'
  import { Image } from './types'

  const props = defineProps({
    image: { type: [String, Object], default: undefined } as Prop<Image>,
    noPhoto: { type: Boolean, default: false },
    noLoader: { type: Boolean, default: false },
    noObserverOnce: { type: Boolean, default: false },
    loadOnce: { type: Boolean, default: false },
    imageHeightFull: { type: Boolean, default: false },
    observerOptions: { type: Object, default: null } as Prop<
      vLazyImgOptions['observerOptions']
    >,
  })

  defineEmits(['intersecting', 'loading', 'loaded', 'error'])

  const sources = computed(() =>
    typeof props.image === 'string'
      ? [{ srcset: props.image }]
      : props.image?.sources,
  )
</script>

<style lang="postcss" scoped>
  /* stylelint-disable no-descending-specificity */
  .m-lazy-img-component {
    @apply maz-relative maz-inline-flex maz-bg-color-lighter maz-flex-center;

    &-loader {
      @apply maz-absolute maz-inset-0 maz-hidden maz-flex-center;
    }

    &:not(.m-lazy-error):not(.m-lazy-no-photo) img {
      @apply maz-h-full maz-w-full;
    }

    &.--height-full img {
      @apply maz-max-h-full maz-w-min maz-max-w-min !important;
    }

    &:not(.m-lazy-loaded):not(.m-lazy-no-photo) {
      & .m-lazy-img-component-loader {
        @apply maz-flex;
      }
    }
  }
</style>
