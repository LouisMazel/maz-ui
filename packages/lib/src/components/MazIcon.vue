<script lang="ts" setup>
import { computed, nextTick, onMounted, ref, watchEffect } from 'vue'

import { useInjectStrict } from '../composables/useInjectStrict'

export interface MazIconProps {
  /** The source path of the SVG file - e.g: `/icons/home.svg` */
  src?: string
  /** The path of the folder where the SVG files are stored - e.g: `/icons` */
  path?: string
  /** The name of the SVG file - e.g: `home` */
  name?: string
  /** The size of the SVG file - e.g: `1em` */
  size?: string
  /** The title of the SVG file - e.g: `Home` */
  title?: string
  /** The function to transform the source of the SVG file - e.g: `(svg) => svg` */
  transformSource?: (svg: SVGElement) => SVGElement
}

const {
  src,
  path,
  name,
  size,
  title,
  transformSource = (svg: SVGElement) => svg,
} = defineProps<MazIconProps>()
const emits = defineEmits<{
  /**
   * emitted when SVG file is loaded
   * @property {SVGElement | undefined} svg the svg element loaded
   */
  (event: 'loaded', svg: SVGElement | undefined): void
  /**
   * emitted when SVG file is not loaded
   */
  (event: 'unloaded'): void
  /**
   * emitted when SVG file is not loaded
   * @property {Error} error the error
   */
  (event: 'error', error: Error): void
}>()
const cache: Record<string, Promise<SVGElement>> = {}
const svgElSource = ref<SVGElement>()
const svgElem = ref<SVGElement>()

function getMazIconPath() {
  try {
    return useInjectStrict<string>('mazIconPath')
  }
  catch {
    return undefined
  }
}

const iconPath = computed(() => path ?? getMazIconPath())
const fullSrc = computed(() => {
  if (src) {
    return src
  }
  else if (iconPath.value) {
    return `${iconPath.value}/${name}.svg`
  }
  else {
    return `/${name}.svg`
  }
})

onMounted(() => {
  if (!name && !src) {
    console.error('[maz-ui](MazIcon) you should provide "name" or "src" as prop')
  }
})

function setTitle(svg: SVGElement, title: string) {
  const titleTags = svg.querySelectorAll('title')
  if (titleTags.length > 0) {
    // overwrite existing title
    titleTags[0].textContent = title
  }
  else {
    // create a title element if one doesn't already exist
    const titleEl = document.createElementNS('http://www.w3.org/2000/svg', 'title')
    titleEl.textContent = title
    svg.append(titleEl)
  }
}

function filterAttrs(attrs: Record<string, unknown>) {
  return Object.keys(attrs).reduce((result, key) => {
    if (attrs[key] !== false && attrs[key] !== null && attrs[key] !== undefined) {
      result[key] = attrs[key]
    }
    return result
  }, {} as Record<string, unknown>)
}

function getSvgAttrs(svgEl: SVGElement) {
  // copy attrs
  const svgAttrs: Record<string, string> = {}
  const attrs = svgEl.attributes
  if (!attrs) {
    return svgAttrs
  }
  for (let i = attrs.length - 1; i >= 0; i--) {
    svgAttrs[attrs[i].name] = attrs[i].value
  }
  return svgAttrs
}

function getSvgContent(svgEl: SVGElement) {
  svgEl.cloneNode(true)
  const svgElNode = transformSource(svgEl)

  if (title) {
    setTitle(svgElNode as SVGElement, title)
  }

  // copy inner html
  return svgEl.innerHTML
}

async function getSource(src: string) {
  // fill cache by src with promise
  if (!cache[src]) {
    // download
    cache[src] = download(src)
  }

  // inline svg when cached promise resolves
  try {
    svgElSource.value = await cache[src]
    // wait to render
    await nextTick()
    emits('loaded', svgElem.value)
  }
  catch (error) {
    if (svgElSource.value) {
      svgElSource.value = undefined
      emits('unloaded')
    }
    // remove cached rejected promise so next image can try load again
    delete cache[src]
    emits('error', error as Error)
  }
}

function download(url: string): Promise<SVGElement> {
  return new Promise((resolve, reject) => {
    const request = new XMLHttpRequest()
    request.open('GET', url, true)

    request.addEventListener('load', () => {
      if (request.status >= 200 && request.status < 400) {
        try {
          // Setup a parser to convert the response to text/xml in order for it to be manipulated and changed
          const parser = new DOMParser()
          const result = parser.parseFromString(request.responseText, 'text/xml')
          let svgEl = result.querySelectorAll('svg')[0] as SVGElement
          if (svgEl) {
            svgEl = transformSource(svgEl)
            resolve(svgEl)
          }
          else {
            reject(new Error('Loaded file is not valid SVG"'))
          }
        }
        catch (error) {
          reject(error)
        }
      }
      else {
        reject(new Error('Error loading SVG'))
      }
    })

    request.addEventListener('error', error => reject(error))
    request.send()
  })
}

watchEffect(() => getSource(fullSrc.value))
</script>

<template>
  <svg
    v-if="svgElSource"
    ref="svgElem"
    class="m-icon m-reset-css"
    width="1em"
    height="1em"
    v-bind="{
      ...getSvgAttrs(svgElSource),
      ...filterAttrs($attrs),
    }"
    :style="`font-size: ${size}`"
    v-html="getSvgContent(svgElSource)"
  />
</template>

<style lang="postcss" scoped>
.m-icon {
  width: 1em !important;
  height: 1em !important;
}
</style>
