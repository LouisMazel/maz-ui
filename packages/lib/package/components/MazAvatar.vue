<template>
  <Component
    :is="componentType"
    :style="{ fontSize: size }"
    class="m-avatar"
    :class="[
      {
        '--has-link': isLink,
      },
    ]"
    v-bind="$attrs"
    :href="href"
    :to="to"
    :target="isLink ? target : undefined"
  >
    <div
      class="m-avatar__wrapper"
      :tabindex="clickable ? 0 : -1"
      :class="{
        '--has-shadow': !noElevation,
        '--bordered': bordered,
        '--clickable': clickable,
        '--square': square,
        '--has-initial': !src && caption,
      }"
      @keydown.enter="clickable ? $emit('click', $event) : undefined"
    >
      <MazLazyImg
        v-if="src"
        class="m-avatar__picture"
        :image="src"
        :alt="alt"
        image-height-full
        :no-loader="noLoader"
        @click="clickable ? $emit('click', $event) : null"
      />
      <slot v-if="caption" name="round-text">
        <span class="m-avatar__initial"> {{ caption?.charAt(0) }} </span>
      </slot>

      <button
        v-if="clickable"
        type="button"
        tabindex="-1"
        class="m-avatar__button maz-flex maz-flex-center"
        @click="$emit('click', $event)"
      >
        <slot name="icon">
          <MazIcon :src="PencilIcon" class="m-avatar__button__icon" />
        </slot>
      </button>
    </div>
    <slot name="caption">
      <p v-if="showCaption && caption" class="m-avatar__caption">
        {{ caption }}
      </p>
    </slot>
  </Component>
</template>

<script lang="ts" setup>
  import { PropType, computed } from 'vue'
  import MazLazyImg from './MazLazyImg.vue'
  import MazIcon from './MazIcon.vue'
  import PencilIcon from './../icons/pencil.svg'

  const props = defineProps({
    // url or path of the image
    src: {
      type: String as PropType<undefined | null | string>,
      default: undefined,
    },
    caption: {
      type: String as PropType<undefined | null | string>,
      default: undefined,
    },
    // url or path to link another page
    href: { type: String, default: undefined },
    // route config
    to: { type: Object, default: undefined },
    // alt text of image
    alt: { type: String, default: 'avatar image' },
    // target attribute of link (if url is provide)
    target: { type: String, default: '_self' },
    // size of avatar
    size: { type: String, default: '1rem' },
    // add border around the avatar
    bordered: { type: Boolean, default: false },
    // add an edit layer & emit `edit` event on click
    clickable: { type: Boolean, default: false },
    // Make the avatar square
    square: { type: Boolean, default: false },
    // Remove the shadow behind the avatar
    noElevation: { type: Boolean, default: false },
    showCaption: { type: Boolean, default: false },
    imageHeightFull: { type: Boolean, default: false },
    noLoader: { type: Boolean, default: false },
  })

  const componentType = computed(() =>
    props.to ? 'RouterLink' : props.href ? 'a' : 'div',
  )
  const isLink = computed(() => !!props.to || !!props.href)

  defineEmits(['click'])
</script>

<style lang="postcss" scoped>
  .m-avatar {
    @apply maz-inline-flex maz-flex-col maz-flex-center;
    @apply maz-no-underline !important;

    &__caption {
      @apply maz-mt-2 maz-w-full maz-truncate maz-text-center maz-text-base maz-font-medium maz-capitalize maz-text-black;
    }

    &__initial {
      @apply maz-capitalize maz-text-white;

      font-size: 1.5em;
    }

    &__wrapper {
      @apply maz-relative maz-flex maz-flex-none maz-justify-center maz-overflow-hidden
      maz-rounded-full maz-bg-color-lighter;

      height: 3em;
      width: 3em;

      &.--clickable {
        & .m-avatar__button {
          @apply maz-outline-none maz-absolute maz-inset-0 maz-w-full maz-cursor-pointer maz-rounded-lg maz-border-none
           maz-bg-transparent maz-opacity-0 maz-transition-all maz-duration-200;

          transform: scale(0);

          &__icon {
            @apply maz-text-white;

            height: 1em;
            width: 1em;
          }
        }

        &:not(.--square) .m-avatar__button {
          @apply maz-rounded-full;
        }

        &:hover,
        &:focus {
          & .m-avatar__picture {
            filter: blur(1.5px);
          }

          & .m-avatar__button {
            @apply maz-opacity-100;

            transform: scale(1.05);
            background-color: var(--maz-color-info-alpha);
          }
        }
      }

      &.--bordered {
        @apply maz-border-2 maz-border-solid maz-border-white;
      }

      &.--square {
        @apply maz-rounded-lg;
      }

      &.--has-shadow {
        @apply maz-shadow;
      }

      &.--has-initial {
        @apply maz-items-center maz-bg-primary;

        &.--clickable {
          & .m-avatar__button {
            @apply maz-bg-danger;
          }
        }
      }
    }

    &.--has-link {
      @apply maz-cursor-pointer;
    }
  }
</style>
