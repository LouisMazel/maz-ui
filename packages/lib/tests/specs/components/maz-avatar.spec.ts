import MazAvatar from '@components/MazAvatar.vue'
import MazLazyImg from '@components/MazLazyImg.vue'
import MazIcon from '@components/MazIcon.vue'
import { shallowMount } from '@vue/test-utils'

describe('MazAvatar', () => {
  test('Component should have expected props', () => {
    const wrapper = shallowMount(MazAvatar)
    const expectedProps = [
      'src',
      'caption',
      'href',
      'to',
      'alt',
      'target',
      'size',
      'bordered',
      'clickable',
      'square',
      'noElevation',
      'showCaption',
      'imageHeightFull',
      'noLoader',
      'buttonColor',
    ]
    expect(Object.keys(wrapper.props())).toEqual(expectedProps)
  })

  test('Component props should have correct default values', () => {
    const wrapper = shallowMount(MazAvatar)
    expect(wrapper.props().src).toBeUndefined()
    expect(wrapper.props().caption).toBeUndefined()
    expect(wrapper.props().href).toBeUndefined()
    expect(wrapper.props().to).toBeUndefined()
    expect(wrapper.props().alt).toBe('avatar image')
    expect(wrapper.props().target).toBe('_self')
    expect(wrapper.props().size).toBeUndefined()
    expect(wrapper.props().bordered).toBe(false)
    expect(wrapper.props().clickable).toBe(false)
    expect(wrapper.props().square).toBe(false)
    expect(wrapper.props().noElevation).toBe(false)
    expect(wrapper.props().showCaption).toBe(false)
    expect(wrapper.props().imageHeightFull).toBe(false)
    expect(wrapper.props().noLoader).toBe(false)
    expect(wrapper.props().buttonColor).toBe('info')
  })

  test('Component should render correctly', () => {
    const wrapper = shallowMount(MazAvatar)
    expect(wrapper.find('.m-avatar').exists()).toBe(true)
    expect(wrapper.find('.m-avatar__wrapper').exists()).toBe(true)
    expect(wrapper.findComponent(MazLazyImg).exists()).toBe(false)
    expect(wrapper.find('.m-avatar__initial').exists()).toBe(false)
    expect(wrapper.find('.m-avatar__button').exists()).toBe(false)
    expect(wrapper.findComponent(MazIcon).exists()).toBe(false)
    expect(wrapper.find('.m-avatar__caption').exists()).toBe(false)
  })
})
