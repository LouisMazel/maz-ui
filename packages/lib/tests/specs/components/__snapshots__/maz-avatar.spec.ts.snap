// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MazAvatar > Should match with the snapshot 1`] = `
"<div data-v-ac4ded86=\\"\\" class=\\"m-avatar\\">
  <div data-v-ac4ded86=\\"\\" class=\\"m-avatar__wrapper --has-shadow\\" tabindex=\\"-1\\">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>"
`;
